sdk/messaging/azservicebus/internal/amqpLinks.go

```diff
@@ -97,240 +101,238 @@ type CreateLinkFunc func(ctx context.Context, session AMQPSession) (AMQPSenderCl
-			tab.StringAttribute("lock", "readlock"),
-			tab.StringAttribute("revisions", fmt.Sprintf("ours(%d), theirs(%d)", ourLinkRevision, *theirLinkRevision)),
-		)
 		return nil
 	}
 
 	links.mu.Lock()
 	defer links.mu.Unlock()
 
-	if theirLinkRevision != nil && ourLinkRevision > *theirLinkRevision {
+	// check once more, just in case someone else modified it before we took
+	// the lock.
+	if links.id.Link != theirLinkRevision.Link {
 		// we've already recovered past their failure.
-		span.AddAttributes(
-			tab.StringAttribute("outcome", "already_recovered"),
-			tab.StringAttribute("lock", "writelock"),
-			tab.StringAttribute("revisions", fmt.Sprintf("ours(%d), theirs(%d)", ourLinkRevision, *theirLinkRevision)),
-		)
 		return nil
 	}
 
-	if err := links.closeWithoutLocking(ctx, false); err != nil {
-		span.Logger().Error(err)
-	}
-
 	err := links.initWithoutLocking(ctx)
 
 	if err != nil {
 		span.AddAttributes(tab.StringAttribute("init_error", err.Error()))
 		return err
 	}
 
-	links.revision++
-
 	span.AddAttributes(
 		tab.StringAttribute("outcome", "recovered"),
-		tab.StringAttribute("revision_new", fmt.Sprintf("%d", links.revision)),
+		tab.StringAttribute("revision_new", fmt.Sprintf("%d", links.id)),
 	)
 	return nil
 }
 
 // Recover will recover the links or the connection, depending
-// on the severity of the error. This function uses the `baseRetrier`
-// defined in the links struct.
-func (links *amqpLinks) RecoverIfNeeded(ctx context.Context, linksRevision uint64, origErr error) error {
+// on the severity of the error.
+func (links *AMQPLinksImpl) RecoverIfNeeded(ctx context.Context, theirID LinkID, origErr error) error {
 	ctx, span := tab.StartSpan(ctx, tracing.SpanRecover)
 	defer span.End()
 
-	var err error = origErr
-
-	retrier := links.baseRetrier.Copy()
-
-	for retrier.Try(ctx) {
-		span.AddAttributes(tab.StringAttribute("recover_attempt", fmt.Sprintf("%d", retrier.CurrentTry())))
-
-		err = links.recoverImpl(ctx, retrier.CurrentTry(), linksRevision, err)
-
-		if err == nil {
-			return nil
-		}
-	}
-
-	return err
-}
-
-func (links *amqpLinks) recoverImpl(ctx context.Context, try int, linksRevision uint64, origErr error) error {
-	_, span := tab.StartSpan(ctx, tracing.SpanRecoverLink)
-	defer span.End()
-
 	if origErr == nil || IsCancelError(origErr) {
 		return nil
 	}
 
+	log.Writef(EventConn, "Recovering link for error %s", origErr.Error())
+
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	default:
 	}
 
-	span.AddAttributes(tab.Int64Attribute("attempt", int64(try)))
-
-	if shouldRecreateLink(origErr) {
-		span.AddAttributes(
-			tab.StringAttribute("recovery_kind", "link"),
-			tab.StringAttribute("error", origErr.Error()),
-			tab.StringAttribute("error_type", fmt.Sprintf("%T", origErr)))
+	sbe := GetSBErrInfo(origErr)
 
-		if err := links.recoverLink(ctx, &linksRevision); err != nil {
-			span.AddAttributes(tab.StringAttribute("recoveryFailure", err.Error()))
+	if sbe.RecoveryKind == RecoveryKindLink {
```

Comment:
Using 'sbe' here instead of the old technique made it simpler to group all of the classification logic into a single spot. It's still all the same logic though, just made a bit simpler by returning constants.